spec/src/main/asciidoc/batch_programming_model.adoc

```diff
@@ -1387,20 +1387,40 @@ parameters.   Note that only the first example using field injection into the 'p
 without an explicit 'name' attribute.
 
 
-===== Consequences And Suggested Patterns
+===== Consequences And Suggested Patterns - Batch Properties
 
 As a consequence of the previous section, an application must be able to get a CDI Bean with a correct view of a batch property by either:
 
 * Injecting the batch property Bean into a @Dependent-scoped CDI Bean via any standard CDI mechanism, (e.g. via a field injection of type: `@Inject @BatchProperty String` within a batch artifact loaded as a CDI Bean).  This assumes the artifact loading will occur on the execution thread.  
    OR
-* Dynamically accessing the batch property bean via `jakarta.enterprise.inject.Instance#get()` or `javax.enterprise.inject.spi.CDI#select()` from the batch execution thread.
+* Dynamically accessing the batch property bean via `jakarta.enterprise.inject.Instance#get()` from the batch execution thread.
```

Comment:
I'm not sure if it is that useful to have instance wrapper there, maybe just "standard cdi injections" is sufficient, instance just bring the ability to check there is  bean but due to jbatch nature this feature is kind of disabled (non binding param).
So my 2cts would be to not reference any type and delegate injections behavior to cdi spec

Reply:
@rmannibucau thanks for your review.   Not sure what you meant exactly by this:
>  instance just bring the ability to check there is bean but due to jbatch nature this feature is kind of disabled (non binding param).

I'm trying to understand if your concern is the fact that we are saying something here or if it is the content of what is being said.

Are you just expressing a preference about the language here, steering us towards a general guideline of leaving it to CDI to describe CDI behavior?

Or are you saying the new example won't necessarily work?   

We have another TCK test, **CDITests#testCDILazyInject**, that tests this exact scenario, using an application-scoped batchlet and verifying that a single injection into the single app-scoped instance will provide the appropriate (different) value when the `Instance.get()` is done in one step vs. the other.  If you think this goes too far then I wonder if you are suggesting we should remove this test as well? 

@rzo1 what is your take on this?

Reply:
@scottkurz my point - not a blocker at all - is that it is not the mainstream CDI pattern and that `Provider` and `Instance` are mainly related to a wider enclosing scope (enclonsing bean is app scoped when value is dependent like in your example) or more advanced cases - outside jbatch. So I don't want to promote that, it can be mentionned as a way to solve that case but default should stay a `@Dependent` bean getting `@Inject @BatchProperty String foo` IMHO. I'd like to avoid to make the integration overcomplex when it is not the case.

Reply:
Hmm, what about converging with something like - please rephrase the english if needed: "Injections under CDI are done using a CDI bean so any CDI injection will work in the context of the injection but JBatch only guarantees the availability of the batch properties values when component is instantiated which means you can need to use an injected `Provider` or `Instance` to lookup lazily a property".

My 2cts would be that it is very rare it would be useful cause very quickly the service bean - app scoped - will be outside batch component instantation and threadlocal setup so let's just not highlight it more than needed.

Reply:
Romain, OK, so it sounds like you are OK with keeping the CDITests#testCDILazyInject test in the TCK.

I'm sorry, I don't follow to the point where I can construct an alternative phrasing from your suggestion.

I would suggest then either:

1. we use my change as-is, removing `or  javax.enterprise.inject.spi.CDI#select()` and adding the new example
2. we remove `or javax.enterprise.inject.spi.CDI#select()` but do NOT add the new example
3. You or someone submit an alternative PR or commit with the exact language proposed

Reply:
Let me note from the history that we did work through this text in detail, back in: 
https://github.com/jakartaee/batch/pull/181#issuecomment-989759326

At the time we had ML discussion, GH issue discussion... we also were prototyping running tests against BatchEE, JBatch, Open Liberty.   We also had a draft of the new, proposed TCK tests.

So I feel like this was somewhat "settled", and the changes should be minimal... only what's necessary to resolve this issue.   

I don't feel like we're overly promoting this pattern either by including it in the spec.  But we're providing justification in case this is ever challenged.

Reply:
@scottkurz well if you showcase in detail this way to do whereas others way are not even mentionned you highlight it and - sadly from experience, not blaming everyone - encourage this pattern more than natural ones.
So long story short either we should include some CDI basis - don't think we should - or just reference CDI as the way to do with the context limitation - which is more a "when" than an "how".
A limitation of the `Instance#get` - or `Provider#get` which is the same - mecanism as showcased is that it leaks memory - creationalcontext to cite it - so a good "dynamic" way is to use `BeanManager` directly and release the creational context once no more needed, this is one example why I think JBatch shouldn't go too far in examples but rather stick to defining in which context the injection is defined.

Reply:
@rmannibucau, so would you then "vote" (informally, I don't mean it's some official vote) for choice 2. from my previous comment? 

> 2. we remove or javax.enterprise.inject.spi.CDI#select() but do NOT add the new example

Thanks

Reply:
My take away is, that the spec shouldn't be altered to much in a minor release, right? If so, maybe we can add @rmannibucau concerns in a future version (if not already discussed for EE11 or EE12) and go with (2) without the example for now?

Reply:
@rzo1 strictly speaking we don't alter the spec at all, we keep 100% the current expected behavior but we explicit a phrasing which was ambiguous.

What about:

> Dynamically accessing the batch property instance via CDI `BeanManager` or  `Provider` (or `Instance`) from the batch execution thread in a component initialization if the bean you want to inject the property in has a scope which survives the component one (`@Dependent` by spec).

Side note: once the components are initialized there is no JBatch guarantee it will work, in particular when threading is involved so guess we can precise this point.

Side note 2: "batch property bean" would be `Bean<String>` and not the `String` value as per CDI semantic so moved to `instance` term too.

No example is very fine for me.

If we want to be more friendly (in terms of completeness) we can add "Please refer to CDI for dynamical lookups to avoid to create memory leak. Releasing the `CreationalContext` is important for example in long running applications - less in standalone batches until the enclosing component is created again and again." but I let this to your judgement.

Reply:
I pushed a new change which simply deletes the reference to `CDI#select()` but does not include the new example.

Reply:
Let me note too that this whole update needs to be reviewed by the Spec Committee.  In general, I don't think Service Releases are supposed to update the spec doc.  That's one more reason to make the text change as small as possible.  I don't know how this will viewed by the spec committee but I'll post a copy to the issue here.

Reply:
Ok for me.

Side note: `Instance#get` does not exist, it is `Provider#get` (https://jakarta.ee/specifications/cdi/4.0/apidocs/jakarta.cdi/jakarta/enterprise/inject/instance for ex), `Instance` just being a `Provider` and you can inject a `Provider` instance (wow this phrasing is likely the most accurate but sounds weird there).

About the spec committee I dont' think there is a blocker since we don't change a single behavior of the spec.

Reply:
OK, the spec committee response is that we shouldn't fix this the spec as part of a 2.1.x service release
(raised https://www.eclipse.org/lists/jakarta.ee-spec/msg03260.html and answered in the same thread with msg 
https://www.eclipse.org/lists/jakarta.ee-spec/msg03282.html).  

Instead we should live with the divergence in the meantime and fix in a future release (e.g. Batch 2.2).  

We can still merge in this PR, though.   I'll also send an email to the Batch ML to make a record of this plan.